File: agent/control/plan_execution.py
"""Plan execution mechanics: approved-plan activation, todo<->step sync, step task
tracking and verification sidechain.

Extracted verbatim from ControlManager (no behavior change). Interaction resolution
(approve / comment / cancel / answer) stays on the facade; this manager holds the
plan-state mechanics those methods drive. Shared queries (`_latest_executable_plan`,
`_has_ask_interaction`) and sub-managers (`permission_tokens`) are reached via `self._cm`.
"""

from __future__ import annotations

from dataclasses import replace
from typing import Any

from ..plans import (
    PlanDraftPhase,
    PlanEvidenceError,
    PlanExecutionState,
    PlanRecord,
    PlanStatus,
    PlanStep,
    PlanStepStatus,
    assess_step_verification,
)
from ..tasks import TaskKind, TaskStatus
from .models import Interaction, now_ts
from .plan_helpers import (
    _is_positive_int,
    _plan_status_from_todo,
    _step_verification_status,
    _task_status_from_plan_step,
)


class PlanExecutionManager:
    def __init__(self, control_manager) -> None:
        self._cm = control_manager

    def sync_plan_from_todos(
        self,
        todos: list[dict[str, Any]],
        *,
        evidence: dict[str, Any] | None = None,
    ) -> PlanRecord | None:
        record = self._cm._latest_executable_plan()
        if record is None or not record.steps:
            return None
        todo_by_step_id = {
            str(item.get("plan_step_id")): item
            for item in todos
            if isinstance(item, dict) and str(item.get("plan_step_id") or "").strip()
        }
        todo_by_index = {
            int(item.get("id")) - 1: item
            for item in todos
            if isinstance(item, dict) and _is_positive_int(item.get("id"))
        }
        now = now_ts()
        steps: list[PlanStep] = []
        for index, step in enumerate(record.steps):
            todo = todo_by_step_id.get(step.id) or todo_by_index.get(index)
            if todo is None:
                steps.append(step)
                continue
            todo_status = str(todo.get("status") or "pending")
            next_status = _plan_status_from_todo(todo_status)
            self._validate_plan_step_transition(step, todo=todo, next_status=next_status)
            step_evidence = list(step.evidence)
            if next_status == PlanStepStatus.DONE.value and step.status != PlanStepStatus.DONE.value:
                step_evidence.append({
                    **(evidence or {}),
                    "todo_id": todo.get("id"),
                    "plan_step_id": todo.get("plan_step_id") or step.id,
                    "todo_status": todo_status,
                    "synced_at": now,
                })
            if next_status == PlanStepStatus.BLOCKED.value and step.status != PlanStepStatus.BLOCKED.value:
                step_evidence.append({
                    **(evidence or {}),
                    "todo_id": todo.get("id"),
                    "plan_step_id": todo.get("plan_step_id") or step.id,
                    "todo_status": todo_status,
                    "blocked_reason": str(todo.get("blocked_reason") or "").strip(),
                    "synced_at": now,
                })
            steps.append(replace(step, status=next_status, evidence=step_evidence))

        plan_status = (
            PlanStatus.COMPLETED.value
            if steps and all(step.status in {PlanStepStatus.DONE.value, PlanStepStatus.SKIPPED.value} for step in steps)
            else PlanStatus.EXECUTING.value
        )
        updated = replace(
            record,
            status=plan_status,
            completed_at=now if plan_status == PlanStatus.COMPLETED.value else record.completed_at,
            updated_at=now,
            steps=steps,
        )
        updated = self._sync_plan_step_tasks(updated)
        self._cm.plan_store.save(updated)
        return updated

    def _validate_plan_step_transition(
        self,
        step: PlanStep,
        *,
        todo: dict[str, Any],
        next_status: str,
    ) -> None:
        if next_status == PlanStepStatus.BLOCKED.value:
            blocked_reason = str(todo.get("blocked_reason") or "").strip()
            if not blocked_reason and not self._cm._has_ask_interaction():
                raise PlanEvidenceError(
                    "PLAN_BLOCKED_REASON_REQUIRED",
                    step_id=step.id,
                    reason="blocked steps must include blocked_reason or be paired with ask_user",
                )
        if next_status != PlanStepStatus.DONE.value or step.status == PlanStepStatus.DONE.value:
            return
        assessment = assess_step_verification(step)
        if assessment.failed_required:
            raise PlanEvidenceError(
                "PLAN_EVIDENCE_FAILED",
                step_id=step.id,
                reason=f"declared verification failed: {'; '.join(assessment.failed_required[:3])}",
            )
        if assessment.blocking_errors:
            raise PlanEvidenceError(
                "PLAN_EVIDENCE_REQUIRED",
                step_id=step.id,
                reason=f"missing passing verification evidence for: {'; '.join(assessment.blocking_errors[:3])}",
            )

    def record_plan_step_tool_output(
        self,
        *,
        tool_name: str,
        summary: str,
        tool_call_id: str | None = None,
        artifacts: list[dict[str, Any]] | None = None,
        metadata: dict[str, Any] | None = None,
        is_error: bool = False,
    ):
        record, step, task_id = self._active_plan_step_task()
        if record is None or step is None or task_id is None or self._cm.task_manager is None:
            return None
        message = {
            "kind": "tool_output",
            "role": "tool",
            "plan_id": record.id,
            "plan_step_id": step.id,
            "tool_name": str(tool_name or ""),
            "tool_call_id": tool_call_id,
            "content": str(summary or "")[:2000],
            "artifacts": artifacts or [],
            "metadata": metadata or {},
            "is_error": bool(is_error),
        }
        self._cm.task_manager.append_sidechain(task_id, message)
        task = self._cm.task_manager.store.get(task_id)
        progress = dict(task.progress) if task is not None else {}
        progress.update({
            "last_tool": str(tool_name or ""),
            "last_summary": str(summary or "")[:500],
            "last_tool_call_id": tool_call_id,
        })
        return self._cm.task_manager.update_task(task_id, progress=progress)

    def _sync_plan_step_tasks(self, record: PlanRecord) -> PlanRecord:
        if self._cm.task_manager is None or not record.steps:
            return record
        mapping = dict(record.metadata.get("plan_step_tasks") or {})
        for index, step in enumerate(record.steps, start=1):
            metadata = {
                "plan_id": record.id,
                "plan_step_id": step.id,
                "sequence": index,
                "verification_status": _step_verification_status(step),
            }
            task_id = str(mapping.get(step.id) or "")
            status = _task_status_from_plan_step(step.status)
            if task_id and self._cm.task_manager.store.get(task_id) is not None:
                task = self._cm.task_manager.store.get(task_id)
                progress = dict(task.progress) if task is not None else {}
                progress["verification_status"] = metadata["verification_status"]
                self._cm.task_manager.update_task(
                    task_id,
                    status=status,
                    title=step.title,
                    metadata=metadata,
                    progress=progress,
                )
                continue
            task = self._cm.task_manager.start_task(
                kind=TaskKind.PLAN_STEP.value,
                title=step.title,
                source="plan_step",
                status=status,
                metadata=metadata,
            )
            mapping[step.id] = task.id
        metadata = dict(record.metadata)
        metadata["plan_step_tasks"] = mapping
        return replace(record, metadata=metadata)

    def _active_plan_step_task(self) -> tuple[PlanRecord | None, PlanStep | None, str | None]:
        record = self._cm._latest_executable_plan()
        if record is None:
            return None, None, None
        mapping = record.metadata.get("plan_step_tasks") or {}
        if not isinstance(mapping, dict):
            return record, None, None
        for step in record.steps:
            if step.status != PlanStepStatus.ACTIVE.value:
                continue
            task_id = str(mapping.get(step.id) or "")
            return record, step, task_id or None
        return record, None, None

    def _append_plan_step_verification(
        self,
        record: PlanRecord,
        *,
        step_id: str,
        result: dict[str, Any],
    ) -> None:
        if self._cm.task_manager is None:
            return
        mapping = record.metadata.get("plan_step_tasks") or {}
        if not isinstance(mapping, dict):
            return
        task_id = str(mapping.get(step_id) or "")
        if not task_id:
            return
        passed = result.get("passed")
        verification_status = "passed" if passed is True else "failed" if passed is False else "unknown"
        self._cm.task_manager.append_sidechain(task_id, {
            "kind": "verification",
            "role": "tool",
            "plan_id": record.id,
            "plan_step_id": step_id,
            "tool_name": str(result.get("source") or "run_command"),
            "command": str(result.get("command") or ""),
            "content": str(result.get("summary") or result.get("error") or "")[:2000],
            "passed": passed,
            "result": dict(result),
        })
        task = self._cm.task_manager.store.get(task_id)
        progress = dict(task.progress) if task is not None else {}
        progress["verification_status"] = verification_status
        progress["last_verification"] = dict(result)
        fields: dict[str, Any] = {"progress": progress}
        if passed is False:
            fields["status"] = TaskStatus.FAILED.value
        self._cm.task_manager.update_task(task_id, **fields)

    def _update_plan_status(self, interaction: Interaction, status: str, *, approved: bool = False) -> None:
        plan_id = str(interaction.meta.get("plan_id") or "")
        if not plan_id:
            return
        record = self._cm.plan_store.get(plan_id)
        if record is None:
            return
        now = now_ts()
        draft = record.draft
        if approved:
            draft = replace(draft, phase=PlanDraftPhase.APPROVED.value)
        payload = {
            **record.to_dict(),
            "status": status,
            "updated_at": now,
            "draft": draft.to_dict(),
        }
        if approved:
            payload["approved_at"] = now
        self._cm.plan_store.save(PlanRecord.from_dict(payload))

    def _activate_approved_plan(self, interaction: Interaction) -> PlanRecord | None:
        plan_id = str(interaction.meta.get("plan_id") or "")
        if not plan_id:
            return None
        record = self._cm.plan_store.get(plan_id)
        if record is None:
            return None
        if self._cm.todo_store is None or not record.steps:
            return record
        activated = PlanExecutionState(record).start_next_step()
        activated = replace(
            activated,
            draft=replace(activated.draft, phase=PlanDraftPhase.EXECUTING.value),
        )
        activated = self._cm.permission_tokens.issue(activated)
        activated = self._sync_plan_step_tasks(activated)
        self._cm.plan_store.save(activated)
        self._cm.todo_store.sync_from_plan_steps([step.to_dict() for step in activated.steps])
        return activated

File: packages/core/src/control/plan-execution.ts
/**
 * PlanExecutionManager (MIG-CTRL-007)。对齐 Python `agent/control/plan_execution.py`。
 * approved→executing 激活、legacy todo→step 投影、step 任务同步、工具输出 sidechain。
 * task_manager 为 null 时（W14 未迁移）所有任务绑定逻辑 no-op —— 与 Python 一致。
 */
import { nowTs } from '../util/time'
import {
  PlanDraftPhase,
  PlanStatus,
  PlanStepStatus,
  planFromDict,
  planToDict,
  type PlanRecord,
  type PlanStep,
} from '../plans/models'
import { PlanExecutionState } from '../plans/execution-state'
import type { Interaction } from './models'
import { plansShareFullGoalScope } from '../goals/scope'
import {
  metadataWithoutPlanPermissionTokens,
  planStatusFromTodo,
  stepVerificationStatus,
  taskStatusFromPlanStep,
} from './plan-helpers'
import type { ControlManagerHost } from './host'

const TASK_KIND_PLAN_STEP = 'plan_step' // agent/tasks TaskKind.PLAN_STEP
const TASK_STATUS_FAILED = 'failed'

export class PlanExecutionManager {
  private readonly cm: ControlManagerHost
  constructor(cm: ControlManagerHost) {
    this.cm = cm
  }

  /**
   * Legacy todo→plan step 投影。Claude Code-style `update_todos` 主链路不调用这里；
   * 保留是为了旧历史、旧测试和显式兼容 API 能读取/投影旧计划状态。
   */
  syncPlanFromTodos(
    todos: Array<Record<string, unknown>>,
    opts?: { evidence?: Record<string, unknown> | null },
  ): PlanRecord | null {
    const record = this.cm.latestExecutablePlan()
    if (record === null || !record.steps.length) return null
    const evidence = opts?.evidence ?? null
    if (record.goalId !== null) {
      const activeGoal = this.cm.activeGoalPlanContext()
      const generation = Number(record.metadata.approval_generation ?? 0)
      if (!activeGoal || activeGoal.id !== record.goalId)
        throw new Error('Goal Todo binding does not match the current Goal')
      for (const item of todos) {
        if (
          String(item.plan_id ?? '').trim() !== record.id ||
          !Number.isInteger(Number(item.approval_generation)) ||
          Number(item.approval_generation) !== generation
        )
          throw new Error(
            'Goal Todo binding must match current plan_id and approval_generation',
          )
      }
    }
    const todoByStepId = new Map<string, Record<string, unknown>>()
    for (const item of todos) {
      if (!item || typeof item !== 'object') continue
      const stepId = String(item.plan_step_id ?? '').trim()
      if (!stepId)
        throw new Error('todo plan_step_id is required for Plan projection')
      if (!record.steps.some((step) => step.id === stepId))
        throw new Error(`unknown todo plan_step_id: ${stepId}`)
      if (todoByStepId.has(stepId))
        throw new Error(`duplicate todo plan_step_id: ${stepId}`)
      todoByStepId.set(stepId, item)
    }
    const now = nowTs()
    let updated = record
    for (const originalStep of record.steps) {
      const todo = todoByStepId.get(originalStep.id)
      if (todo === undefined) continue
      const todoStatus = String(todo.status ?? 'pending')
      const nextStatus = planStatusFromTodo(todoStatus)
      const currentStep = updated.steps.find(
        (step) => step.id === originalStep.id,
      )!
      if (currentStep.status === nextStatus) continue
      const transitionEvidence = {
        ...(evidence ?? {}),
        todo_id: todo.id,
        plan_step_id: originalStep.id,
        todo_status: todoStatus,
        ...(nextStatus === PlanStepStatus.BLOCKED
          ? { blocked_reason: String(todo.blocked_reason ?? '').trim() }
          : {}),
        synced_at: now,
      }
      if (
        (nextStatus === PlanStepStatus.ACTIVE ||
          nextStatus === PlanStepStatus.DONE) &&
        currentStep.status === PlanStepStatus.PENDING
      ) {
        updated = new PlanExecutionState(updated).startNextStep()
      }
      const active = updated.steps.find((step) => step.id === originalStep.id)!
      if (nextStatus === PlanStepStatus.DONE) {
        updated = new PlanExecutionState(updated).completeStep(
          originalStep.id,
          {
            evidence: transitionEvidence,
          },
        )
      } else if (nextStatus === PlanStepStatus.BLOCKED) {
        updated = new PlanExecutionState(updated).blockStep(originalStep.id, {
          evidence: transitionEvidence,
        })
      } else if (
        nextStatus === PlanStepStatus.ACTIVE &&
        active.status !== PlanStepStatus.ACTIVE
      ) {
        throw new Error(
          `plan step ${originalStep.id} dependencies are not satisfied`,
        )
      } else if (nextStatus === PlanStepStatus.PENDING) {
        throw new Error('todo projection cannot move a Plan step backwards')
      }
    }
    updated = { ...updated, updatedAt: now }
    updated = this.syncPlanStepTasks(updated)
    return this.cm.planStore.save(updated)
  }

  /** 批准新计划时取代同 store 内滞留的 approved/executing 旧计划，防止僵尸累积（B1）。 */
  supersedeStaleExecutingPlans(newPlanId: string): void {
    const successor = this.cm.planStore.get(newPlanId)
    if (successor === null) throw new Error('successor Plan does not exist')
    const now = nowTs()
    for (const record of this.cm.planStore.list()) {
      if (record.id === newPlanId) continue
      if (
        record.status !== PlanStatus.APPROVED &&
        record.status !== PlanStatus.EXECUTING
      )
        continue
      if (
        record.goalId !== successor.goalId ||
        !plansShareFullGoalScope(record, successor)
      )
        continue
      const taskMap =
        record.metadata.plan_step_tasks &&
        typeof record.metadata.plan_step_tasks === 'object' &&
        !Array.isArray(record.metadata.plan_step_tasks)
          ? {
              ...(record.metadata.plan_step_tasks as Record<string, string>),
            }
          : {}
      const metadata = metadataWithoutPlanPermissionTokens(record.metadata, {
        reason: 'Plan superseded by an approved successor',
      })
      metadata.plan_step_tasks_revoked = taskMap
      metadata.plan_step_tasks = {}
      metadata.superseded_by = newPlanId
      metadata.superseded_at = now
      metadata.supersession_audit = {
        predecessor_plan_id: record.id,
        successor_plan_id: newPlanId,
        goal_id: successor.goalId,
      }
      // 不碰 updatedAt：取代是记账而非活动，latest() 必须继续指向新计划
      this.cm.planStore.save({
        ...record,
        status: PlanStatus.CANCELLED,
        metadata,
      })
      for (const taskId of Object.values(taskMap))
        this.cm.taskManager?.cancelTask(taskId, {
          reason: 'Plan superseded by an approved successor',
        })
    }
  }

  recordPlanStepToolOutput(opts: {
    toolName: string
    summary: string
    toolCallId?: string | null
    artifacts?: Array<Record<string, unknown>> | null
    metadata?: Record<string, unknown> | null
    isError?: boolean
  }): unknown {
    const [record, step, taskId] = this.activePlanStepTask()
    if (
      record === null ||
      step === null ||
      taskId === null ||
      this.cm.taskManager === null
    )
      return null
    const message = {
      kind: 'tool_output',
      role: 'tool',
      plan_id: record.id,
      plan_step_id: step.id,
      tool_name: String(opts.toolName ?? ''),
      tool_call_id: opts.toolCallId ?? null,
      content: String(opts.summary ?? '').slice(0, 2000),
      artifacts: opts.artifacts ?? [],
      metadata: opts.metadata ?? {},
      is_error: Boolean(opts.isError),
    }
    this.cm.taskManager.appendSidechain(taskId, message)
    const task = this.cm.taskManager.store.get(taskId)
    const progress = task !== null ? { ...task.progress } : {}
    progress.last_tool = String(opts.toolName ?? '')
    progress.last_summary = String(opts.summary ?? '').slice(0, 500)
    progress.last_tool_call_id = opts.toolCallId ?? null
    return this.cm.taskManager.updateTask(taskId, { progress })
  }

  private syncPlanStepTasks(record: PlanRecord): PlanRecord {
    if (this.cm.taskManager === null || !record.steps.length) return record
    const mapping = {
      ...((record.metadata.plan_step_tasks as Record<string, string>) ?? {}),
    }
    const scope = planStepTaskScope(record, this.cm.planScopeMetadata())
    record.steps.forEach((step, idx) => {
      const index = idx + 1
      const metadata = {
        plan_id: record.id,
        plan_step_id: step.id,
        approval_generation: Number(record.metadata.approval_generation ?? 0),
        sequence: index,
        verification_status: stepVerificationStatus(step),
        ...(scope ? { scope } : {}),
      }
      const taskId = String(mapping[step.id] ?? '')
      const status = taskStatusFromPlanStep(step.status)
      if (taskId && this.cm.taskManager!.store.get(taskId) !== null) {
        const task = this.cm.taskManager!.store.get(taskId)
        const progress = task !== null ? { ...task.progress } : {}
        progress.verification_status = metadata.verification_status
        this.cm.taskManager!.updateTask(taskId, {
          status,
          title: step.title,
          metadata,
          progress,
        })
        return
      }
      const task = this.cm.taskManager!.startTask({
        kind: TASK_KIND_PLAN_STEP,
        title: step.title,
        source: 'plan_step',
        status,
        sessionId:
          record.sessionId ??
          (this.cm.planScopeMetadata()?.session_id as string | undefined) ??
          null,
        metadata,
      })
      mapping[step.id] = task.id
    })
    const metadata = { ...record.metadata }
    metadata.plan_step_tasks = mapping
    return { ...record, metadata }
  }

  private activePlanStepTask(): [
    PlanRecord | null,
    PlanStep | null,
    string | null,
  ] {
    const record = this.cm.latestExecutablePlan()
    if (record === null) return [null, null, null]
    const mapping = record.metadata.plan_step_tasks
    if (!mapping || typeof mapping !== 'object') return [record, null, null]
    for (const step of record.steps) {
      if (step.status !== PlanStepStatus.ACTIVE) continue
      const taskId = String((mapping as Record<string, string>)[step.id] ?? '')
      return [record, step, taskId || null]
    }
    return [record, null, null]
  }

  appendPlanStepVerification(
    record: PlanRecord,
    opts: { stepId: string; result: Record<string, unknown> },
  ): void {
    if (this.cm.taskManager === null) return
    const mapping = record.metadata.plan_step_tasks
    if (!mapping || typeof mapping !== 'object') return
    const taskId = String(
      (mapping as Record<string, string>)[opts.stepId] ?? '',
    )
    if (!taskId) return
    const passed = opts.result.passed
    const verificationStatus =
      passed === true ? 'passed' : passed === false ? 'failed' : 'unknown'
    this.cm.taskManager.appendSidechain(taskId, {
      kind: 'verification',
      role: 'tool',
      plan_id: record.id,
      plan_step_id: opts.stepId,
      tool_name: String(opts.result.source ?? 'run_command'),
      command: String(opts.result.command ?? ''),
      content: String(opts.result.summary ?? opts.result.error ?? '').slice(
        0,
        2000,
      ),
      passed,
      result: { ...opts.result },
    })
    const task = this.cm.taskManager.store.get(taskId)
    const progress = task !== null ? { ...task.progress } : {}
    progress.verification_status = verificationStatus
    progress.last_verification = { ...opts.result }
    const fields: Record<string, unknown> = { progress }
    if (passed === false) fields.status = TASK_STATUS_FAILED
    this.cm.taskManager.updateTask(taskId, fields)
  }

  updatePlanStatus(
    interaction: Interaction,
    status: string,
    opts?: { approved?: boolean },
  ): PlanRecord | null {
    const planId = String(interaction.meta.plan_id ?? '')
    if (!planId) return null
    const record = this.cm.planStore.get(planId)
    if (record === null) return null
    const now = nowTs()
    let draft = record.draft
    if (opts?.approved) draft = { ...draft, phase: PlanDraftPhase.APPROVED }
    const scope = this.cm.planScopeMetadata()
    const metadata = {
      ...record.metadata,
      ...(scope && !record.metadata.scope ? { scope } : {}),
    }
    const payload: Record<string, unknown> = {
      ...planToDict(record),
      status,
      updated_at: now,
      draft: {
        ...(planToDict({ ...record, draft }).draft as Record<string, unknown>),
      },
      metadata,
    }
    if (opts?.approved) payload.approved_at = now
    return this.cm.planStore.save(planFromDict(payload))
  }

  activateApprovedPlan(interaction: Interaction): PlanRecord | null {
    const planId = String(interaction.meta.plan_id ?? '')
    if (!planId) return null
    const record = this.cm.planStore.get(planId)
    if (record === null) return null
    if (this.cm.todoStore === null || !record.steps.length) return record
    let activated = new PlanExecutionState(record).startNextStep()
    activated = {
      ...activated,
      draft: { ...activated.draft, phase: PlanDraftPhase.EXECUTING },
    }
    activated = this.cm.permissionTokens.issue(activated)
    activated = this.syncPlanStepTasks(activated)
    return this.cm.planStore.save(activated)
  }
}

function planStepTaskScope(
  record: PlanRecord,
  current: Record<string, unknown> | null,
): Record<string, unknown> | null {
  const saved = record.metadata.scope
  if (saved && typeof saved === 'object' && !Array.isArray(saved))
    return { ...(saved as Record<string, unknown>) }
  return current ? { ...current } : null
}
